base/reflection.jl

```diff
@@ -1239,11 +1239,65 @@ true
 ```
 """
 function hasmethod(@nospecialize(f), @nospecialize(t); world=typemax(UInt))
+    hasmethod(f, to_tuple_type(t); world=world)
+end
+
+# This is just declaring a simplified version of `@generated` earlier in bootstrappping
+# to be used by hasmethod. TODO: Replace this with just using @eval ?
+macro _early_generated(headsig, body)
+    return Expr(:escape,
+                Expr(:function, headsig,
+                     Expr(:block,
+                          Expr(:if, Expr(:generated),
+                               body,
+                               Expr(:block,
+                                    Expr(:meta, :generated_only),
+                                    Expr(:return, nothing))))))
+end
+
+# This is used to create the CodeInfo returned by hasmethod in the case of false.
+_hasmethod_false(@nospecialize(f), @nospecialize(t), @nospecialize(w)) = false
+
+@_early_generated(
+    static_hasmethod(
+        @nospecialize(f),
+        @nospecialize(t::Type{T}),
+        @nospecialize(world::Val{W})
+    ) where {T<:Tuple, W},
+    begin
+        fi = f.instance  #TODO: make this work with constructors and functors.
+        typ = signature_type(fi, T)
+        method_doesnot_exist = ccall(:jl_gf_invoke_lookup, Any, (Any, UInt), typ, W) === nothing
+        if method_doesnot_exist
+            ci_orig = uncompressed_ast(typeof(_hasmethod_false).name.mt.defs.func)
+            ci = ccall(:jl_copy_code_info, Ref{CodeInfo}, (Any,), ci_orig)
+
+            # Now we add the edges so if a method is defined this recompiles
+            mt = f.name.mt
+            ci.edges = Core.Compiler.vect(mt, typ)
+            return ci
+        else
+            return true  # We are done, it exists, no need to recompile ever
+            # except if it is deleted. TODO: deal with Base.delete_method
+        end
+    end
+)
+
+function static_hasmethod(@nospecialize(f), @nospecialize(t); world=typemax(UInt))
+    return static_hasmethod(f, t, Val{world}())
+end
+
+
+function dynamic_hasmethod(@nospecialize(f), @nospecialize(t); world=typemax(UInt))
     t = to_tuple_type(t)
     t = signature_type(f, t)
     return ccall(:jl_gf_invoke_lookup, Any, (Any, UInt), t, world) !== nothing
 end
 
+function hasmethod(@nospecialize(f), @nospecialize(t); world=typemax(UInt))
```

Comment:
This signature seems identical to the one above, aren't you getting an overwrite warning?

Reply:
good catch.